### PR TITLE
Merge XML report output

### PR DESCRIPTION
**Pull request by chao2zhang**
This is a rewrite of #3394 and addresses one item in #3360.

- Only the gradle tasks are exposed: No opinionated way to configure the tasks.
- Documentation is added.
- Fix: `utf-8` to `UTF-8`, because the latter is the [standard](https://www.w3schools.com/xml/xml_syntax.asp).

### CODE REVIEW

File: docs/pages/reporting.md [hunk 1]
```diff
@@ -71,6 +71,61 @@ detekt {
...
+
+#### Kotlin DSL
+
+```kotlin
+val xmlReportMerge by tasks.registering(io.gitlab.arturbosch.detekt.report.XmlReportMergeTask::class) { 
+    output.set(project.layout.buildDirectory.file("reports/detekt/merge.xml"))
+}
+
+subprojects {
+    detekt {
```
Comment (cortinico) [inline docs/pages/reporting.md:114]: This block should be 2 space indented

File: detekt-gradle-plugin/src/intTest/kotlin/io/gitlab/arturbosch/detekt/DetektXmlReportMergeTest.kt [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+                    addSubmodule(
+                        name = "child2",
+                        numberOfSourceFilesPerSourceDir = 4,
+                        numberOfCodeSmells = 4
+                    )
+                }
+
+                val gradleRunner = DslGradleRunner(projectLayout, builder.gradleBuildName, mainBuildFileContent)
+                gradleRunner.setupProject()
+                gradleRunner.runTasksAndExpectFailure("detekt", "xmlReportMerge", "--continue") { result ->
```
Comment (cortinico) [inline detekt-gradle-plugin/src/intTest/kotlin/io/gitlab/arturbosch/detekt/DetektXmlReportMergeTest.kt:93]:
```suggestion
                gradleRunner.runTasksAndExpectFailure("detekt", "xmlReportMerge", "--continue", "--stacktrace") { result ->
```

That's a nit. Also I'm unsure you need the `--continue`.

File: detekt-gradle-plugin/src/main/kotlin/io/gitlab/arturbosch/detekt/report/XmlReportMergeTask.kt [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    @get:InputFiles
+    @get:PathSensitive(PathSensitivity.RELATIVE)
+    abstract val input: ConfigurableFileCollection
+
+    @get:OutputFile
+    abstract val output: RegularFileProperty
+
+    @TaskAction
+    fun merge() {
+        logger.lifecycle(input.files.joinToString(separator = "\n") { it.absolutePath })
```
Comment (cortinico) [inline detekt-gradle-plugin/src/main/kotlin/io/gitlab/arturbosch/detekt/report/XmlReportMergeTask.kt:25]:
```suggestion
        logger.info(input.files.joinToString(separator = "\n") { it.absolutePath })
```

File: detekt-gradle-plugin/src/main/kotlin/io/gitlab/arturbosch/detekt/report/XmlReportMergeTask.kt [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+        XmlReportMerger.merge(input.files, output.get().asFile)
+        logger.lifecycle("Merged XML output to ${output.get().asFile.absolutePath}")
```
Comment (cortinico) [inline detekt-gradle-plugin/src/main/kotlin/io/gitlab/arturbosch/detekt/report/XmlReportMergeTask.kt:27]: Also this should go to the `info`? But probably one line is ok to live on the `lifecycle` logger.

**Review by cortinico — APPROVED** [review]

File: detekt-gradle-plugin/src/intTest/kotlin/io/gitlab/arturbosch/detekt/DetektXmlReportMergeTest.kt (hunk 2 above)
Comment (chao2zhang) [reply detekt-gradle-plugin/src/intTest/kotlin/io/gitlab/arturbosch/detekt/DetektXmlReportMergeTest.kt:93]: Ah good call. I have left my debugging code here.

File: detekt-gradle-plugin/src/main/kotlin/io/gitlab/arturbosch/detekt/report/XmlReportMergeTask.kt (hunk 3 above)
Comment (chao2zhang) [reply detekt-gradle-plugin/src/main/kotlin/io/gitlab/arturbosch/detekt/report/XmlReportMergeTask.kt:25]: Good call. I have promoted this to lifecycle for debugging.

File: detekt-gradle-plugin/src/main/kotlin/io/gitlab/arturbosch/detekt/report/XmlReportMergeTask.kt (hunk 4 above)
Comment (chao2zhang) [reply detekt-gradle-plugin/src/main/kotlin/io/gitlab/arturbosch/detekt/report/XmlReportMergeTask.kt:27]: I would prefer this merged XML output to be lifecycle so people can find the link in the console log.

File: docs/pages/reporting.md (hunk 1 above)
Comment (chao2zhang) [reply docs/pages/reporting.md]: Fixed

File: detekt-gradle-plugin/src/intTest/kotlin/io/gitlab/arturbosch/detekt/DetektXmlReportMergeTest.kt (hunk 2 above)
Comment (schalkms) [inline detekt-gradle-plugin/src/intTest/kotlin/io/gitlab/arturbosch/detekt/DetektXmlReportMergeTest.kt:93]: I'm afraid, I don't quite understand, why we expect a failure for the configured task.

**Review by schalkms — COMMENTED** [review]
I really like this feature. I had a short look at it. It works fine. Great! 👍

I think it's good to start small with the XML report format. For me XML reports supporting a merge is the first logical step. Other reports could follow later on, if this is highly valuable.

I'm not sure regarding the test coverage for the XML parsing logic and task. Should we include more edge cases?

**Review by BraisGabin — APPROVED** [review]

File: detekt-gradle-plugin/src/intTest/kotlin/io/gitlab/arturbosch/detekt/DetektXmlReportMergeTest.kt (hunk 2 above)
Comment (chao2zhang) [reply detekt-gradle-plugin/src/intTest/kotlin/io/gitlab/arturbosch/detekt/DetektXmlReportMergeTest.kt:93]: Because the `detekt` task should fail due to detected issues. `xmlReportMerge` should succeed.

**Comment by chao2zhang** [comment]
> I really like this feature. I had a short look at it. It works fine. Great! 👍
> 
> I think it's good to start small with the XML report format. For me XML reports supporting a merge is the first logical step. Other reports could follow later on, if this is highly valuable.
> 
> I'm not sure regarding the test coverage for the XML parsing logic and task. Should we include more edge cases?

The logic is relatively straightforward so I think the existing test coverage should be fine and jacoco is happy about hitting the 80% coverage in this diff.
